src/digest.rs

```diff
@@ -0,0 +1,119 @@
+use serde::Serialize;
+use sha3::{Digest, Sha3_256};
+use std::iter::Extend;
+use std::marker::PhantomData;
+
+use crate::constants::NUM_HASH_BITS;
+
+/// For building digests
+#[derive(Clone)]
+pub struct DigestBuilder<F: PrimeField, T: HasDigest<F>> {
+  bytes: Vec<u8>,
+  hasher: Option<Sha3_256>,
+  bytes_digest: [u8; 32],
+  inner: Option<T>,
+  _p: PhantomData<(F, T)>,
+}
+
+/// Trait to be implemented by types whose digests can be built with `DigestBuilder`.
+pub trait HasDigest<F: PrimeField> {
+  /// Extend `bytes` with raw bytes or digest summarizing `Digestible`.
+  fn set_digest(&mut self, digest: F);
+  fn set_bytes_digest(&mut self, _bytes_digest: [u8; 32]) {
+    unimplemented!()
+  }
+}
+
+/// Trait for components with potentially discrete digests to be included in their container's digest.
+pub trait Digestible {
+  /// Extend a byte sink with the bytes to be digested.
+  fn extend_bytes<X: Extend<u8>>(&self, bytes: &mut X);
+}
+
+/// Marker trait to be implemented for types that implement `Digestible` and `Serialize`.
+/// Their instances will be serialized to bytes then digested.
+pub trait SimpleDigestible: Serialize {}
+
+impl<T: SimpleDigestible> Digestible for T {
+  fn extend_bytes<X: Extend<u8>>(&self, bytes: &mut X) {
+    (*bytes).extend(bincode::serialize(self).unwrap());
+  }
+}
+
+impl<F: PrimeField, T: HasDigest<F> + Digestible> DigestBuilder<F, T> {
+  /// Return a new `DigestBuilder`.
+  pub fn new() -> Self {
+    Self {
+      bytes: Default::default(),
+      hasher: Some(Sha3_256::new()),
+      bytes_digest: [0; 32],
+      inner: None,
+      _p: Default::default(),
+    }
+  }
+
+  /// Initialize `DigestBuilder` with `inner`.
+  pub fn init(&mut self, inner: T) {
+    self.inner = Some(inner)
+  }
+
+  /// Update builder's hasher.
+  pub fn update(&mut self) {
+    if let Some(hasher) = self.hasher.as_mut() {
+      hasher.update(&self.bytes);
+    } else {
+      panic!("hasher missing");
+    };
+
+    self.bytes.truncate(0);
+  }
+
+  pub fn finalize_bytes_digest(&mut self) {
+    self.update();
+
+    let Some(hasher) = self.hasher.take() else {panic!("hasher missing");};
+    self.bytes_digest = hasher.finalize().into();
+  }
+
+  /// Finalize digest.
+  pub fn finalize(&mut self) -> F {
+    self.finalize_bytes_digest();
+
+    let digest = self.bytes_digest;
+    let bv = (0..NUM_HASH_BITS).map(|i| {
+      let (byte_pos, bit_pos) = (i / 8, i % 8);
+      let bit = (digest[byte_pos] >> bit_pos) & 1;
+      bit == 1
+    });
+
+    // turn the bit vector into a scalar
+    let mut digest = F::ZERO;
+    let mut coeff = F::ONE;
+    for bit in bv {
+      if bit {
+        digest += coeff;
+      }
+      coeff += coeff;
+    }
+    digest
+  }
+
+  /// Extend `DigestBuilder` with the bytes provided by the underlying inner `HasDigest`.
+  pub fn compute_digest(&mut self) -> F {
+    let o = self.inner.as_ref().expect("inner Digestible missing");
+    o.extend_bytes(&mut self.bytes);
```

Comment:
It seems we write into `self.bytes` as a buffer and then immediately consume it with `finalize()`. However, there's no other place where `self.bytes` can be extended other than `new()`, which initializes it as empty. Can't this just be replaced locally with:
```rust
let mut bytes = vec![];
o.extend_bytes(&mut bytes);
```
Is `self.bytes` designing for future use case where we are extending `bytes` and updating the hasher dynamically?

Reply:
Yes, the inner object may need to implement a more interesting (than just serializing itself) method of providing bytes. For example, if it contains other `Digestible` things (like circuits), then it will want to just provide its own digest — which may already have been computed. We don't necessarily need to actually accumulate all these bytes before hashing. Probably just updating the hasher itself will be fine (rather than saving the bytes and doing all the hashing last). I left it as-is to stay close to the previous work until we know exactly what the total process will be. It may be that changing to eliminate the accumulated bytes will have some effect on the digest itself because of potential padding considerations. My goal for now was to refactor without changing any digest results, though.